### PR TITLE
Deprecate default conversions

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #87006

This switches off the path for default conversions - from now on, only explicitly registered conversions work.

```release-note
k8s.io/apimachinery - scheme.Convert() now uses only explicitly registered conversions - default reflection based conversion is no longer available. `+k8s:conversion-gen` tags can be used with the `k8s.io/code-generator` component to generate conversions.
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/conversion/converter.go [hunk 1]
```diff
@@ -442,8 +442,27 @@ func (c *Converter) doConversion(src, dest interface{}, flags FieldMatchingFlags
 	if fn, ok := c.generatedConversionFuncs.untyped[pair]; ok {
 		return fn(src, dest, scope)
 	}
-	// TODO: consider everything past this point deprecated - we want to support only point to point top level
-	// conversions
+
+	if pair.source == pair.dest {
+		// TODO: Ideally, to avoid this hack, we should auto-generate self-conversions.
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/conversion/converter.go:447]: include the EnforcePtr(src) and EnforcePtr(dest) checks here?

Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/conversion/converter.go:447]: done

File: staging/src/k8s.io/apimachinery/pkg/conversion/converter.go [hunk 2]
```diff
@@ -442,8 +442,33 @@ func (c *Converter) doConversion(src, dest interface{}, flags FieldMatchingFlags
 	if fn, ok := c.generatedConversionFuncs.untyped[pair]; ok {
 		return fn(src, dest, scope)
 	}
-	// TODO: consider everything past this point deprecated - we want to support only point to point top level
-	// conversions
+
+	if pair.source == pair.dest {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/conversion/converter.go:446]: would it be better or worse to register a generated self-conversion function when we add an object to a scheme in AddKnownTypeWithName, like:

```go
	// if the type implements DeepCopyInto(<obj>), register a self-conversion
	if m := reflect.ValueOf(obj).MethodByName("DeepCopyInto"); !m.IsZero() && m.Type().NumIn() == 1 && m.Type().NumOut() == 0 && m.Type().In(0) == reflect.TypeOf(obj) {
		if err := s.AddGeneratedConversionFunc(obj, obj, func(a, b interface{}, scope conversion.Scope) error {
			// copy a to b
			reflect.ValueOf(a).MethodByName("DeepCopyInto").Call([]reflect.Value{reflect.ValueOf(b)})
			// clear typemeta to match legacy reflective conversion
			b.(Object).GetObjectKind().SetGroupVersionKind(schema.GroupVersionKind{})
			return nil
		}); err != nil {
			panic(err)
		}
	}
```

that would remove all reflective conversion from this method, and would let do `SetGroupVersionKind(schema.GroupVersionKind{})` to match legacy behavior

File: staging/src/k8s.io/apimachinery/pkg/conversion/converter.go [hunk 3]
```diff
@@ -442,8 +442,33 @@ func (c *Converter) doConversion(src, dest interface{}, flags FieldMatchingFlags
...
+			if deepCopyIntoType.NumIn() == 1 && deepCopyIntoType.In(0) == pair.dest && deepCopyIntoType.NumOut() == 0 {
+				reflect.ValueOf(src).MethodByName(deepCopyIntoName).Call([]reflect.Value{reflect.ValueOf(dest)})
+				// TODO: Conversions were cleaning up TypeMeta field, which DeepCopy is copying.
+				//   It seems, it was causing only one unit test to failed, so for now we don't clean it up.
+				return nil
+			}
+		}
+	}
+
+	return fmt.Errorf("converting (%s) to (%s): unknown conversion", pair.source, pair.dest)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/conversion/converter.go:468]: nit: trim the `*` prefix from pair.source.String() and pair.dest.String() to keep the error messages identical?

File: staging/src/k8s.io/apimachinery/pkg/conversion/converter.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/conversion/converter.go:446]: great idea - let me try this

File: staging/src/k8s.io/apimachinery/pkg/conversion/converter.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/conversion/converter.go]: Done (though in slightly different way, by keeping the preexisting EnforcePtr stuff).

File: staging/src/k8s.io/apimachinery/pkg/conversion/converter.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/conversion/converter.go]: Done - seems to work great. Thanks a lot for this suggestion - this is much cleaner now.

**Comment by deads2k** [comment]
/hold

I'd like a chance to have a look at this and judge impact.  I expect to have an answer back this week.  The PR itself looks pretty contained, but it's relying on a fair number of 1.18 PRs so it will take me a bit.

**Comment by deads2k** [comment]
Well, it's not great, but it looks tractable.  Looks like a dozen unique bits or so https://github.com/openshift/openshift-apiserver/pull/95

/hold cancel

**Comment by liggitt** [comment]
this needs a release note for `k8s.io/apimachinery` use, pointing at conversion-gen or something similar

**Comment by wojtek-t** [comment]
> this needs a release note for k8s.io/apimachinery use, pointing at conversion-gen or something similar

@liggitt - Added something - PTAL

**Pull request merged**